### ISSUE TITLE
How to implement DTOs avoiding an explicit dependency to TrackerDog

### ISSUE THREAD

**Issue by mfidemraizer** [mfidemraizer/trackerdog#21]
@gtmacgregor has raised a question in issue #19:

> I want to have a lightweight readonly version of the DTO which does not rely on external dependencies (TrackerDog) so to do this I have an interface which is readonly (no setters). The implementation of that interface contains internal setters for each property. Then I created a mutable implementation which derives from the readonly version exposing those internal setters as public. If I only had one implementation which was read write then I could not force the end users of the library to get a tracked version to make changes.

I've moved here to be able to close #19 and discuss his concerns here :)

**Comment by mfidemraizer** [comment]
@gtmacgregor Uhm, while I understand what you said, I'm trying to figure out what's the requirement behind your concerns.

Let me know if I guessed right your problem:

- You want DTOs to be **read-only** because you want to return them and avoid the risk of someone setting them while this should be done in a lower layer (i.e. domain). 

- You're implementing a library which would work as I've mentioned in the previous point, but also you want that consumers of that library could get a *mutable* version of these DTOs.

Am I mistaken?

**Comment by gtmacgregor** [comment]
@mfidemraizer 

Thanks for all the time helping!

Yes you are correct, I would like to have a set of read-only DTO's for the reason you mentioned as well as reducing dependencies, don't want trackerdog as a dependency on the read-only versions as those will be consumed by many who have no need to change anything.  So due to that requirement I do need an analogous set of objects which are read-write in which I want to force change tracking so they can be properly sent to the data store. Currently I have a set of interfaces and implementations which allow me to do this but it does not give the granularity I need to do efficient change tracking like I would get with trackerdog. Also this method requires I have a read-only and mutable version of each class which I would like to avoid...although I am not sure I can because the property setters are hidden.

My 2 goals are:
1. To have a set of DTO's that can't be changed or at least not changed without those changes being tracked 
2. Ensure all changes are tracked so those can be sent to datastore.

I hope this clears up some questions.

**Comment by gtmacgregor** [comment]
I am thinking of using a generic decorator to make the read-only objects mutable using a SetProperty function but that will require reflection which I don't really want to employ if possible.

**Comment by mfidemraizer** [comment]
@gtmacgregor 

I believe that your issue is you're creating trackable objects inside class constructors (as I checked in the other issue #19).

Maybe you need to design your DTOs as follows:

```
public class CustomerDto : IReadOnlyCustomerDto
{
       public string Name { get; set; }
}

public interface IReadOnlyCustomerDto 
{
      public string Name { get; }
}
```

DTOs are read-write by default, but they implement *read-only* interfaces. When you need to transfer from data to a tier where you don't want read-only DTOs, you just need to return them casted to `IReadOnlyWhatever`:

```
public IList<IReadOnlyCustomerDto> GetCustomerList()
{
        List<CustomerDto> customers = new List<CustomerDto>();
        // stuff...
        
        return customers.Cast<IReadOnlyCustomerDto>().ToList(); 
}
```

Or, if you're fine with `IEnumerable<T>`, since `T` is covariant...

```
public IEnumerable<IReadOnlyCustomerDto> GetCustomerList()
{
        List<CustomerDto> customers = new List<CustomerDto>();
        // stuff...
        
        // Implicit cast
        return customers; 
}
```

**So, where to turn DTOs into change-trackable?**. I don't know how you taylored your application or service, but I guess you've some domain layer/business layer and there you can turn DTOs into change-trackable when needed.

And **how you get read-only non-tracked versions of your DTOs and as `IReadOnlyWhatever`?** Easy, you can use **[AutoMapper ](http://automapper.org/)** for this:

```
public IReadOnlyCustomerDto GetReadOnlyCustomer(Guid id)
{
         CustomerDto dto = .... // who knows how you get this DTO :D
        
         // Lets turn it into a read-only DTO without tracking:
        return Mapper.Map<CustomerDto, CustomerDto>(dto);
}
```

[AutoMapper ](http://automapper.org/) can map an existing DTO of a given type to a new instance of the same type. That is, you get a non-tracked DTO which can be implicitly casted to `IReadOnlyWhatever`.

Does this solve your concerns?

**Comment by gtmacgregor** [comment]
@mfidemraizer Thank you!

I have another question, so what if my CustomerDto contains a list of ContactDto's. Ideally in the readonly interface I would expose IEnumerable<IContact> (IContact is readonly) so my IReadOnlyCustomerDto is still readonly by virtue of the IEnumerable and IContact...can't add/remove anything is IEnumerable and IContact is the readonly interface.

I can make another tracked property in the implementation of CustomerDto, something like IList<Contact> but then when I am using the tracked object the original interface function on IReadOnlyCustomerDto is always empty as it is not tracked. That, to me is confusing and will be to the end users as well.

Any thoughts?

Thanks

 Glenn

**Comment by gtmacgregor** [comment]
@mfidemraizer My goal is to just force the user to use a tracked version of the object if they are going to make changes so those changes can be sent to the data store. Maybe I am going about it wrong or it is not possible.

Thanks

 Glenn

**Comment by mfidemraizer** [comment]
@gtmacgregor  

First of all, I understand that the `IReadOnlyWhatever` approach is fine for you?

## About the next question/concern

Don't use `IEnumerable<T>` to represent read-only collections. [Use `System.Collections.Immutable`](https://msdn.microsoft.com/en-us/library/system.collections.immutable(v=vs.111).aspx) and an [`IImmutableList<T>`](https://msdn.microsoft.com/en-us/library/dn467185(v=vs.111).aspx).

Did you know that you can implement the same property from different interfaces with [**explicit implementations**](https://msdn.microsoft.com/en-us/library/ms173157.aspx)? Explicit implementations will be only available if the whole object is casted to the interface implemented explicitly!

Your `IReadOnlyCustomerDto`  would define the contact list as follows:

```
public interface IReadOnlyCustomerDto 
{
        IImmutableList<IReadOnlyContactDto> Contacts { get; }
}
```

And your `CustomerDto` should look as follows:

```
public class CustomerDto : IReadOnlyCustomerDto
{
         // This is the change-trackable list
         public IList<ContactDto> Contacts { get; set; } = new List<ContactDto>();

         // IReadOnlyCustomerDto exposes an immutable list containing fresh IReadOnlyContactDto DTOs
         IImmutableList<IReadOnlyContactDto> IReadOnlyCustomerDto.Contacts => Contacts.Cast<IReadOnlyContactDto>().ToImmutableList();
}
```

```
CustomerDto dto = new CustomerDto();
dto.Contacts // <--- this will access the tracked contact list

IReadOnlyCustomerDto readOnlyDto = dto;
readOnlyDto.Contacts // <-- this is the immutable list!
```

**Comment by gtmacgregor** [comment]
@mfidemraizer Thank you, working well!

I am having an issue with getting collection change info. I tried to follow the examples to get the IReadOnlyChangeTrackableCollection by:

IReadOnlyChangeTrackableCollection c = (IReadOnlyChangeTrackableCollection)entity.Contacts; 

Contacts returns an IList<Contact> but I get a runtime error when the above is executed: 

Unable to cast object of type 'System.Collections.Generic.List`1[Contact]' to type 'TrackerDog.IReadOnlyChangeTrackableCollection'.

An I missing something?

Thanks

 Glenn

**Comment by gtmacgregor** [comment]
@mfidemraizer 

I did some digging and found in your tests when you create a trackable WhateverParent from an existing WhateverParent in CollectionItemsArePreservedWhenTurningParentObjectIntoTrackable the list are also proxied. Meaning before the object is trackable List and List2 are just Lists but after the CreateFrom the tracked object List and List2 types are now IList Proxy. This is not happening for my object:

`
Customer {
private IList<Contact> mContacts = new List<Contact>();
...
public virtual IList<Contact> Contacts 
{
   get { return mContacts;}
   set mContacts = value.ToList();}
}
`

So when I create a trackable Customer the type of Contacts is still the list and not the IList proxy. All the objects in the list are proxy objects as I have configured that property to be tracked as well as the Contact object.

Thanks

 Glenn
